data_structures/src/superblock.rs

```diff
@@ -544,6 +570,39 @@ fn magic_partition<T: Clone>(v: &[T], first: usize, size: usize) -> Vec<T> {
     v_subset
 }
 
+// Take size element out of v.len() starting with element at index first plus an offset:
+// magic_partition(v, 3, 3, r), v=[0, 1, 2, 3, 4, 5], r=[1].
+// Will return elements at index 4, 0, 2.
+fn magic_partition_2<T>(v: &[T], first: usize, size: usize, rand_distribution: &[u8]) -> HashSet<T>
+where
+    T: Clone + Eq + std::hash::Hash,
+{
+    let mut hs_subset = HashSet::default();
+    if first >= v.len() {
+        return hs_subset;
+    }
+    let each = v.len() / size;
+
+    let mut step_index = 0_usize;
+    let mut step = rand_distribution[step_index] as usize;
+    let mut a = first;
+    let mut b = (a + step) % v.len();
+    while hs_subset.len() < size {
```

Comment:
This could be an infinite loop if `v.len() < size`, but currently that's not possible because the size if checked before calling this function. But I guess we can add an `assert!(v.len() >= size)` to be safe.